tsl/src/hypercore/hypercore_handler.c

```diff
@@ -1787,6 +1789,87 @@ hypercore_tuple_delete(Relation relation, ItemPointer tid, CommandId cid, Snapsh
 	return result;
 }
 
+/*
+ * Decompress a segment that contains the row given by ctid.
+ *
+ * This function is typically called during an upsert (ON CONFLICT DO UPDATE),
+ * where the conflicting row points to a compressed segment that needs to be
+ * decompressed before the update can take place. This function is used to
+ * decompress that segment into a set of individual rows.
+ *
+ * Returns the number of rows in the segment that were decompressed, or 0 if
+ * the TID pointed to a regular (non-compressed) tuple. The TID of the
+ * de-compressed conflicting row is returned via "new_ctid".
+ */
+int
+hypercore_decompress_conflict_segment(Relation relation, const ItemPointer ctid,
+									  TupleTableSlot *slot, Snapshot snapshot, ItemPointer new_ctid)
+{
+	HypercoreInfo *hcinfo;
+	Relation crel;
+	TupleTableSlot *cslot;
+	ItemPointerData decoded_tid;
+	TM_Result result;
+	TM_FailureData tmfd;
+	int n_batch_rows = 0;
+	uint16 tuple_index;
+	bool should_free;
+
+	/* Nothing to do if this is not a compressed segment */
+	if (!is_compressed_tid(ctid))
+		return 0;
+
+	Assert(TTS_IS_ARROWTUPLE(slot));
+
+	hcinfo = RelationGetHypercoreInfo(relation);
+	crel = table_open(hcinfo->compressed_relid, RowExclusiveLock);
+	tuple_index = hypercore_tid_decode(&decoded_tid, ctid);
+
+	/* Is it guaranteed that "slot" is populated with the conflicting segment
+	 * tuple?  Otherwise we need to fetch the compressed segment... Unclear if
+	 * this is ever needed. */
+	if (TTS_EMPTY(slot) && !table_tuple_fetch_row_version(relation, ctid, snapshot, slot))
```

Comment:
I am not 100% sure this fetch is needed, hence the comment above.

Reply:
It looks like it is guaranteed.

1. Before calling this function in `ExecOnConflictUpdate`, `table_tuple_lock` will be called with the TID of the conflicting tuple.
2. Since you have a check for `!is_compressed_tid(ctid)` above, we only need to consider the `is_compressed_tid` path in `hypercore_tuple_lock`.
3. This branch in `hypercore_tuple_lock` fetches the child slot for the slot passed in and calls `heapam_tuple_lock` on the compressed relation with the TID for the compressed tuple part of the conflicting TID.
4. `heapam_tuple_lock` does a lot of voodoo magic to fetch the heap tuple, but ends with `ExecStorePinnedBufferHeapTuple(tuple, slot, buffer)` which stores the locked tuple in the slot.
5. Later in `ExecOnConflictUpdate` this function is called, which will then contain the compressed tuple with the conflicting tuple.

It might be prudent to keep this check to deal with changes upstreams, but we have this comment in `tableam.h`:

```c
 * Lock a tuple in the specified mode.
    .
    .
    .
 * Output parameters:
 *	*slot: contains the target tuple
 *	*tmfd: filled in failure cases (see below)
```

So it seems that there is a guarantee that the slot will contain the target tuple, hence also the compressed tuple with the conflict.